py-polars/polars/internals/frame.py

```diff
@@ -536,7 +536,7 @@ def read_parquet(
         if columns:
             if isinstance(columns, list):
                 if all(isinstance(i, int) for i in columns):
-                    projection = columns  # type: ignore
+                    projection = tp.cast(tp.List[int], columns)
```

Comment:
just as a suggestion (feel free to ignore), if you don't want to use `cast`, there's `TypeGuard` in `typing_extensions`. This looks very similar to the example from the [docs](https://www.python.org/dev/peps/pep-0647/#typeguard-type), here it could be
```python
def _is_int_list(val: List[object]) -> TypeGuard[List[int]]:
    return all(isinstance(x, int) for x in val)
```
, then with
```python
if _is_int_list(columns):
    projection = columns
```
should just work

Reply:
That is the more elegant solution indeed. I did have to remove the `[object]`, otherwise mypy was not satisfied. Not sure why not.

Reply:
> Not sure why not.

Because `list` is invariant, meaning that although `int` is `object` it doesn't follow that `List[int]` is `List[object]` (however, `Iterable` is covariant, so there it does hold that `Iterable[int]` is `Iterable[object]`):
```python
import typing as tp

a: tp.List[object]
b: tp.List[int]
b = a

c: tp.Iterable[object]
d: tp.Iterable[int]
c = d
```

I think it should work to annotate `projection` as `Optional[Iterable[int]]`, and to change `is_str_list(val: List) -> TypeGuard[List[str]]` to `is_str_iterable(val: Iterable[object]) -> TypeGuard[Iterable[str]]`

Reply:
See below, I have opted for Sequence instead of Iterable.